### PR TITLE
Feature/dont use regex for url parsing

### PR DESCRIPTION
**Pull request by carlosesilva**
Reimplement the url parsing functionality using the DOMDocument class instead of regex.

This pull request also has some other improvements compared to #2 .
- It saves the urls keyed by tag name, which means we only need to keep one post meta entry per post. (all the post meta stuff is being handled in #3, which will be getting updated with this code once this pr is merged)
- It searches for insecure content in additional tags such as responsive images using srcset and using picture tags with source tags inside them.
- Makes it simple to update the tags to search for insecure content by keeping an array of tags instead of one giant regex pattern.

### CODE REVIEW

**Comment by carlosesilva** [comment]
Ooops, I just saw that I need to update the unit tests since I have changed things around

File: bu-ssl.php [hunk 1]
```diff
@@ -242,34 +293,106 @@ public function remove_all_postmeta() {
...
+			// Query the content for the specified tag.
+			$dom_node_list = $dom->getElementsByTagName( $tag['name'] );
 
-			// If the url is internal, ignore it.
-			if ( 0 === strpos( $u[2], get_site_url( null, null, 'http' ) ) ) {
-				array_splice( $urls, $k, 1 );
+			// Initilize dom_node_list_array which will hold all of the DOMNodeList's we will check for insecure content.
+			$dom_node_list_array = array();
+
+			// If $tag['children'] is present.
```
Comment (Adam42) [inline bu-ssl.php]: Can probably combine these two comments into one. "Get any children's DOMNodeList's from the base DOMNodeList".

File: bu-ssl.php [hunk 2]
```diff
@@ -297,22 +420,30 @@ public function has_insecure_content( $content = '', $search_type = 'any' ) {
...
+
+				// Merge insecure image urls.
+				$insecure_images = array_unique( array_merge( $insecure_imgs, $insecure_pictures ) );
 
-			if ( ! empty( $urls ) ) {
-				// Create a proxy url and replace the insecure url with it.
-				foreach ( $urls as $k => $u ) {
-					$url = $u[2];
-					$content = str_replace( $url, $camo->proxy( $url ), $content );
+				// If there are insecure images.
```
Comment (Adam42) [inline bu-ssl.php]: Can remove `// If there are insecure images.`, the conditional is pretty clear.

File: bu-ssl.php [hunk 3]
```diff
@@ -78,10 +78,61 @@ class SSL {
...
 		'content_security_policy'   => "default-src https: 'unsafe-inline' 'unsafe-eval'",
 		'csp_report_url'            => '',
-		// regex adopted from @imme_emosol https://mathiasbynens.be/demo/url-regex .
-		'http_img_regex'            => '@<img.*src\s{0,4}=.{0,4}(http:\/\/[^\s\/$.?#].[^\s\'"]*).+>@iS',
-		// see full list -- https://developer.mozilla.org/en-US/docs/Web/Security/Mixed_content#Types_of_mixed_content .
-		'http_all_regex'            => '@<(img|audio|video|object|iframe|script|link|iframe).*(?:src|data|href)\s{0,4}=.{0,4}(http:\/\/[^\s\/$.?#].[^\s\'"]*).+>@iS',
+	);
+
+	/**
+	 * The html tags that can have insecure content
```
Comment (Adam42) [inline bu-ssl.php]: Should be html tags that might have insecure content we'll want to proxy or warn about. Can implies that they're allowed insecure content.

File: bu-ssl.php [hunk 4]
```diff
@@ -242,34 +293,106 @@ public function remove_all_postmeta() {
...
 			}
+
+			// Iterate over all DOMNodeList's.
+			foreach ( $dom_node_list_array as $dom_node_list ) {
+				// Iterate over all elements in DOMNodeList.
+				foreach ( $dom_node_list as $element ) {
+					// Get the url from the specified attribute.
+					$attribute_value = $element->getAttribute( $tag['attribute'] );
+
+					// If attribute_value is not empty.
```
Comment (Adam42) [inline bu-ssl.php]: Superfluous comment could be removed? Comments should be for code that needs explaining or to call attention to something important or odd, IMO.

File: bu-ssl.php [hunk 5]
```diff
@@ -242,34 +293,106 @@ public function remove_all_postmeta() {
...
+					if ( '' !== $attribute_value ) {
+
+						$element_urls = array();
+						if ( 'srcset' === $tag['attribute'] ) {
+							$element_urls = $this->get_urls_from_srcset_attribute( $attribute_value );
+						} else {
+							$element_urls = array( $attribute_value );
+						}
+
+						// Iterate over element_urls.
```
Comment (Adam42) [inline bu-ssl.php]: Superfluous comment

File: bu-ssl.php [hunk 6]
```diff
@@ -242,34 +293,106 @@ public function remove_all_postmeta() {
...
+						foreach ( $element_urls as $url ) {
+							// Parse url using the wp_parse_url function.
```
Comment (Adam42) [inline bu-ssl.php]: Also a superfluous comment.

File: bu-ssl.php [hunk 7]
```diff
@@ -297,22 +420,30 @@ public function has_insecure_content( $content = '', $search_type = 'any' ) {
...
 			$camo->setCamoKey( BU_SSL_CAMO_KEY );
 
-			// Get list of insecure img urls.
-			$urls = self::has_insecure_content( $content, 'img' );
+			// Get list of insecure urls.
+			$insecure_urls_per_tag = self::has_insecure_content( $content );
+			if ( $insecure_urls_per_tag ) {
+
+				// Get insecure urls from img and picture tags.
+				$insecure_imgs = $insecure_urls_per_tag['img'] ? $insecure_urls_per_tag['img'] : array();
```
Comment (Adam42) [inline bu-ssl.php]: Should the first portion of the ternary be wrapped in `isset`? e.g. `isset($insecure_urls_per_tag['img']) ? $insecure_urls_per_tag['img'] : array();` and same for 'picture' below?

**Review by Adam42 — APPROVED** [review]
Approved and looks good and a great change overall but note comments about unnecessary comments and most critically this question about `isset`:

https://github.com/bu-ist/bu-ssl/pull/4/files#r206757476

Would like for @anton-kachurin or @jdub233 to weigh in prior to merging.

**Comment by jdub233** [comment]
I took a look and overall this looks like a good change to me.  I'm very supportive of using DOMDocument instead of regex for HTML markup parsing.

I agree with Adam that some of the comments are a bit redundant.  Also I'm starting to wonder if we shouldn't be looking to re-write insecure links instead of recording and proxying them, but that is almost certainly a discussion outside this PR.

I haven't tested the changes, but overall this looks good to me.

**Comment by Adam42** [comment]
I agree we should have a discussion on re-writing insecure links instead of proxying but also agree that is well outside the scope of this PR. But, recording will let us see how many of the insecure URLs are to outside content that doesn't have a secure alternative so we'd have to proxy those anyway!

**Pull request merged**